### PR TITLE
Guidelines pre-PEP: Subset of C, some Types, Return values, Output arguments

### PR DESCRIPTION
**Pull request by encukou**
It's been a while since I [last proposed some guidelines](https://github.com/capi-workgroup/api-evolution/pull/51). Here's the next chunk.

@zooba @vstinner @erlend-aasland @serhiy-storchaka @mdboom, please add your comments or approvals.

Like last time, if a section is controversial I can delete it for now, and split it out into a separate discussion.
Smaller issues that need substantial extra work/discussion can go in the checklist at https://github.com/capi-workgroup/api-evolution/issues/44.

### CODE REVIEW

File: guidelines.rst [hunk 1]
```diff
@@ -282,9 +282,104 @@ that is, the types of all arguments must be specified.
...
+
+Avoid ``enum``, which have compiler-/platform-dependent size.
+(CPython can not yet use C23's fixed underlying `enum` types.)
+Instead, use ``int`` with defined constants.
+
+Also avoid bitfields, which have compiler-/platform-dependent memory layout.
+Instead, use fixed width integer types and bitmask constants.
+
+To be clear: it is fine to use ``enum`` and bitfields outside the
+public headers.
```
Comment (zooba) [inline guidelines.rst:381]:
```suggestion
To be clear: it is fine to use ``enum`` and bitfields within CPython sources,
provided they do not need to be used in public headers.
```

File: guidelines.rst [hunk 2]
```diff
@@ -309,3 +404,193 @@ These objects should be type-checked as if they were ``PyObject*``.
...
+for example:
+
+* ``-1`` or ``NULL`` is a valid (non-exceptional) result: for example,
+  in ``PyDict_GetItemRef``.
+* There are multiple results: for example, in ``PyUnicode_AsUTF8AndSize``.
+
+Guidelines for output arguments:
+
+* Functions must always fill in the output arguments. If an error
+  occurs or the rseult is not available, the output should typically be set
```
Comment (zooba) [inline guidelines.rst:499]:
```suggestion
  occurs or the result is not available, the output should typically be set
```

File: guidelines.rst [hunk 3]
```diff
@@ -282,9 +282,104 @@ that is, the types of all arguments must be specified.
...
+will be either more performant, or easier to use from C.
+For example:
+
+*  A function may be shadowed by a ``static inline`` function or macro with
+   the same behavior. Typically, this allows better performance for C/C++.
+   See `shadowing example`_.
+*  A function that uses C-specific types, such as ``PyLong_AsLongLong``,
+   is OK if equivalent functions are provided for the preferred types.
+*  A variadic function is OK if there's an equivalent function
+   that takes an array.
```
[the remark this answers is not in the document]
Comment (serhiy-storchaka) [reply guidelines.rst:312]:
```
PyObject* PyObject_Call(PyObject *callable, PyObject *args, PyObject *kwargs)
PyObject* PyObject_CallFunction(PyObject *callable, const char *format, ...)
```
I do not see array here.

The closest analogues are
```
PyObject* PyObject_Vectorcall(PyObject *callable, PyObject *const *args, size_t nargsf, PyObject *kwnames)
PyObject* PyObject_VectorcallDict(PyObject *callable, PyObject *const *args, size_t nargsf, PyObject *kwdict)
```
And this (with addition of `PyObject_VectorcallMethod`) are only examples of such functions.

[the remark this answers is not in the document]
Comment (zooba) [reply guidelines.rst:312]: `PyObject *args` is a tuple of arguments, which is an array-like alternative to the varargs in `CallFunction`.

We don't mean they must literally be a raw pointer to a sequential list of values. But there should be a reasonably efficient way for non-C code to call the function.

[the remark this answers is not in the document]
Comment (serhiy-storchaka) [reply guidelines.rst:312]: I think that this is too wide interpretation of term "array", especially in the context of the C API. Could we use something less misleading?

[the remark this answers is not in the document]
Comment (zooba) [reply guidelines.rst:312]: I'd be happy with "...an equivalent function that doesn't use varargs but allows the caller to pass all values in a single call, for example with an array". That might be a bit wordy though.

File: guidelines.rst [hunk 4]
```diff
@@ -309,3 +404,193 @@ These objects should be type-checked as if they were ``PyObject*``.
...
+
+    static inline returntype
+    _Py_Foo_impl(ARGS)
+    {
+        ...
+    }
+
+    PyAPI_FUNC(returntype) Py_Foo (ARGS);
+
+    #define Py_Foo _Py_Foo_impl
```
Comment (mdboom) [inline guidelines.rst:538]: Some functions will want to expose the library function in the stable ABI for stable ABI reasons (at the expense of performance).  How is that normally done -- some sort of difference based on `Py_LIMITED_API`?  I realize that's a different concept, but it still might be good to have an example of a function exposed through the limited API, but exposed as static inline or macro through the non-limited API.

**Review by mdboom — APPROVED** [review]

File: guidelines.rst [hunk 5]
```diff
@@ -282,9 +282,104 @@ that is, the types of all arguments must be specified.
...
+Once you add API that conforms to this “portable subset”,
+you can add additional C/C++-specific API. Usually, the additional API
+will be either more performant, or easier to use from C.
+For example:
+
+*  A function may be shadowed by a ``static inline`` function or macro with
+   the same behavior. Typically, this allows better performance for C/C++.
+   See `shadowing example`_.
+*  A function that uses C-specific types, such as ``PyLong_AsLongLong``,
+   is OK if equivalent functions are provided for the preferred types.
```
Comment (erlend-aasland) [inline guidelines.rst:310]:
```suggestion
   is OK if equivalent functions are provided for :ref:`the preferred types <types>`.
```

File: guidelines.rst [hunk 6]
```diff
@@ -282,9 +282,104 @@ that is, the types of all arguments must be specified.
...
+
+
+Enums and bitfields
+-------------------
+
+Avoid ``enum``, which have compiler-/platform-dependent size.
+(CPython can not yet use C23's fixed underlying `enum` types.)
+Instead, use ``int`` with defined constants.
+
+Also avoid bitfields, which have compiler-/platform-dependent memory layout.
```
Comment (erlend-aasland) [inline guidelines.rst:377]:
```suggestion
Avoid bitfields, which have compiler-/platform-dependent memory layout.
```

File: guidelines.rst [hunk 7]
```diff
@@ -309,3 +404,193 @@ These objects should be type-checked as if they were ``PyObject*``.
     - https://github.com/capi-workgroup/api-evolution/issues/29
     - https://github.com/capi-workgroup/decisions/issues/19
 
+
+Return values
+=============
+
+The return value of a function must indicate whether an exception was set.
+It must not be necessary to use ``PyErr_Occurred`` to disambiguate.
```
Comment (erlend-aasland) [inline guidelines.rst:412]: Should we emphasise this more?

File: guidelines.rst [hunk 8]
```diff
@@ -309,3 +404,193 @@ These objects should be type-checked as if they were ``PyObject*``.
...
+
+* ``-1`` or ``NULL`` is a valid (non-exceptional) result: for example,
+  in ``PyDict_GetItemRef``.
+* There are multiple results: for example, in ``PyUnicode_AsUTF8AndSize``.
+
+Guidelines for output arguments:
+
+* Functions must always fill in the output arguments. If an error
+  occurs or the rseult is not available, the output should typically be set
+  to NULL or zero.
```
Comment (erlend-aasland) [inline guidelines.rst:500]:
```suggestion
  to ``NULL`` or zero.
```

File: guidelines.rst [hunk 9]
```diff
@@ -309,3 +404,193 @@ These objects should be type-checked as if they were ``PyObject*``.
...
+* When it might be useful for users to call a function but ignore an output,
+  allow passing NULL as the output argument.
```
Comment (erlend-aasland) [inline guidelines.rst:502]:
```suggestion
  allow passing ``NULL`` as the output argument.
```

File: guidelines.rst [hunk 10]
```diff
@@ -309,3 +404,193 @@ These objects should be type-checked as if they were ``PyObject*``.
...
+* Ownership of a ``PyObject*`` result is transferred to the caller,
+  as with return values.
+  [TODO: Link to guidelines about ownership & borrowing, when those are added]
```
Comment (erlend-aasland) [inline guidelines.rst:505]:
```suggestion

    .. TODO: Link to guidelines about ownership & borrowing, when those are added
```

File: guidelines.rst [hunk 11]
```diff
@@ -309,3 +404,193 @@ These objects should be type-checked as if they were ``PyObject*``.
...
+
+*  Success or failure: return ``int``
+
+   * ``0`` for success
+   * ``-1``, with an exception set, for failure
+
+*  Yes or no: return ``int``
+
+   *  ``0`` for ``false``
+   *  ``1`` for ``true``
```
Comment (erlend-aasland) [inline guidelines.rst:570]: Nit; match the description (yes before no) and visually separate the 1 and -1 case by arranging them 1, 0, -1:

```suggestion
   *  ``1`` for ``true``
   *  ``0`` for ``false``
```

File: guidelines.rst [hunk 12]
```diff
@@ -309,3 +404,193 @@ These objects should be type-checked as if they were ``PyObject*``.
...
+   *  ``-1``, with an exception set, for failure
+
+*  Lookup (“getattr”, “getitem” or “setdefault” style) functions: return
+   ``int``; the lookup result is passed via an
+   :ref:`output argument <output argument>`):
+
+   *  ``0`` for “not found”
+      (*result* is set to ``NULL`` or other zero/empty value)
+   *  ``1`` for “found” (*result* is set)
```
Comment (erlend-aasland) [inline guidelines.rst:579]: Ditto.

```suggestion
   *  ``1`` for “found” (*result* is set)
   *  ``0`` for “not found”
      (*result* is set to ``NULL`` or other zero/empty value)
```

File: guidelines.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (vstinner) [reply guidelines.rst:312]: When PySys_Audit() was added to the limited C API 3.13, `int PySys_AuditTuple(const char *event, PyObject *args)` was also added: pass an array of Python objects as a tuple.

File: guidelines.rst [hunk 13]
```diff
@@ -282,9 +282,104 @@ that is, the types of all arguments must be specified.
...
+As always, new exceptions can be added with approval from the C API
+working group.
+
+.. note::
+
+    For background and discussions, see:
+
+    - https://github.com/capi-workgroup/api-evolution/issues/11 (Treat the ABI as an API)
+    - https://github.com/capi-workgroup/api-evolution/issues/18 (Avoid macros and static inline functions)
+    - https://github.com/capi-workgroup/api-evolution/issues/12 (variadics)
```
[the remark this answers is not in the document]
Comment (encukou) [reply guidelines.rst:339]: This is ReST, not Markdown.
I'll leave this kind of formatting for later, when it's time to turn this into a PEP.
But, “background and discussions” is a great place for the PEP 670 link:
```suggestion
    - :pep:`670`
    - https://github.com/capi-workgroup/api-evolution/issues/11 (Treat the ABI as an API)
    - https://github.com/capi-workgroup/api-evolution/issues/18 (Avoid macros and static inline functions)
    - https://github.com/capi-workgroup/api-evolution/issues/12 (variadics)
```

File: guidelines.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (encukou) [reply guidelines.rst:312]: Keep in mind the WG may grant exceptions (esp. for consistency with existing API).

This is already a list of examples, but let's repeat that?
```suggestion
*  A variadic function is OK if there's an equivalent function
   that, for example, takes an array or tuple.
```

FWIW, if `PyUnicode_FromFormat` was added today, I *would* argue for a version that takes an array (of tagged unions).

File: guidelines.rst [hunk 14]
```diff
@@ -282,9 +282,104 @@ that is, the types of all arguments must be specified.
...
+*  ``int32_t`` and other C99+ fixed width integer types
+*  ``Py_ssize_t``, ``intptr_t``, ``ptrdiff_t`` for values of the appropriate
+   platform-specific types
+*  ``char``
+*  ``double`` (IEEE 754 ``binary64``)
+
+The only exception is ``int``, which should be used for small ranges
+(typically, as a replacement for enum).
+If the 16-bit limit is relevant, and for all unsigned values, prefer
+explicit fixed width types over ``int``.
```
Comment (encukou) [inline guidelines.rst:364]: > char can be signed or not, I would suggest to not use it in new APIs for arithmetic types.

Right. But, it's useful for strings.

```suggestion
*  ``int32_t`` and other C99+ fixed width integer types
*  ``Py_ssize_t``, ``intptr_t``, ``ptrdiff_t`` for values of the appropriate
   platform-specific types
*  ``double`` (IEEE 754 ``binary64``)

As an exception, use ``int`` for small ranges (typically, as a replacement
for enum).
If the 16-bit limit is relevant, and for all unsigned values, prefer
explicit fixed width types over ``int``.

As another exception, use ``char*`` for UTF-8/ASCII strings (and ``char``
for bytes of such text). For byte-strings, prefer ``uint8_t*`` or ``void*``.
```

File: guidelines.rst (hunk 7 above)
Comment (encukou) [reply guidelines.rst:412]: I don't think so; the second redundant sentence should be fine.
Since it's a principle that guides the other guidelines, the reader will be fine if they skip it.

File: guidelines.rst (hunk 10 above)
Comment (encukou) [reply guidelines.rst:505]: I need to do a formatting pass before turning this into a PEP. Before that, I prefer visible TODO reminders.

File: guidelines.rst (hunk 4 above)
Comment (encukou) [reply guidelines.rst:538]: Do this, but wrap the `static inline` function and the `#define` line in `#if !defined(Py_LIMITED_API)`.

I don't immediately know where to put this, so I'll [save your comment for later](https://github.com/capi-workgroup/api-evolution/issues/44).

**Review by vstinner — APPROVED** [review]
LGTM

File: guidelines.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (serhiy-storchaka) [reply guidelines.rst:312]: `PySys_Audit()` and `PySys_AuditTuple()` is a good example. `PySys_Audit() = Py_BuildValue() + PySys_AuditTuple()`.

Other examples are:
`PyErr_Format() = PyUnicode_FromFormat() + PyErr_SetObject()`
`PyErr_WarnFormat() = PyUnicode_FromFormat() + ... + PyErr_WarnExplicit()`
`PyErr_WarnExplicitFormat() = PyUnicode_FromFormat() + PyErr_WarnExplicit()`
`PyUnicodeWriter_Format() = PyUnicode_FromFormat() + PyUnicodeWriter_WriteStr()`

There equivalent functions take a string. So I would add a string to example.

Array cannot be used for heterogeneous arguments. Functions like `PyUnicode_FromFormat()` are C-specific. Other programming languages can implement equivalent function using native means or do not implement them at all.

**Review by erlend-aasland — APPROVED** [review]

File: guidelines.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (encukou) [reply guidelines.rst:312]: What about:
```suggestion
*  A variadic function is OK if there's an non-variadic function
   with equivalent functionality.
   Usually, the equivalent take an array or pre-constructed
   Python object: for example, `PyObject_Vectorcall` is the equivalent
   for `PyObject_CallFunction`; `PySys_AuditTuple` for `PySys_Audit`;
   `PyErr_SetObject` for `PyErr_Format`.
```

> Other programming languages can implement equivalent function using native means or do not implement them at all.

Yes. What's important that it's *possible* (and better, *easy*) to implement them by other means. 
For string formatting, `PyUnicodeWriter` helps :)

[the remark this answers is not in the document]
Comment (encukou) [reply guidelines.rst:312]: I agree, but the example is already long. It doesn't need to be complete :)

(Most of the `PyObject_Call*` functions are equivalent to one another -- i.e. redundant -- but that's a [different issue](https://github.com/capi-workgroup/api-evolution/issues/48).)

**Comment by encukou** [comment]
Serhiy, are you OK with this version?

(I'll treat earlier reviews as valid: The changes were minor, and everyone will get plenty of time to re-review when the document is complete.)

**Review by serhiy-storchaka — APPROVED** [review]
LGTM. Thanks @encukou.
